// wdio-ui5-service/dist/service.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
exports.__esModule = true;
var multiremote_1 = require("webdriverio/build/multiremote");
var wdi5_bridge_1 = require("./lib/wdi5-bridge");
var Logger_1 = require("./lib/Logger");
var Logger = Logger_1.Logger.getInstance();
var Service = /** @class */ (function () {
    function Service(_options, // TODO: this is the successor to _config in wdio^8
    _capabilities, _config // an enhanced version of the regular wdio config
    ) {
        this._options = _options;
        this._capabilities = _capabilities;
        this._config = _config;
    } // the Service is instantiated by wdio with the capabilites and config passed on to
    Service.prototype.before = function ( /*capabilities* , specs*/) {
        return __awaiter(this, void 0, void 0, function () {
            var timeout, _a, _b, _c, _d, _e, _f, _g, name_1, e_1_1;
            var e_1, _h;
            return __generator(this, function (_j) {
                switch (_j.label) {
                    case 0:
                        // if no wdi5 config is available we add it manually
                        if (!this._config.wdi5) {
                            this._config["wdi5"] = {};
                        }
                        return [4 /*yield*/, (0, wdi5_bridge_1.start)(this._config)];
                    case 1:
                        _j.sent();
                        Logger.info("started");
                        return [4 /*yield*/, (0, wdi5_bridge_1.setup)(this._config)];
                    case 2:
                        _j.sent();
                        Logger.info("setup complete");
                        timeout = (this._config["wdi5"]["waitForUI5Timeout"] || 15000) + 5000;
                        return [4 /*yield*/, browser.setTimeout({ script: timeout })];
                    case 3:
                        _j.sent();
                        Logger.debug("browser script timeout set to ".concat(timeout));
                        if (!(typeof browser.getTimeouts === "function")) return [3 /*break*/, 5];
                        _b = (_a = Logger).debug;
                        _c = "browser timeouts are ".concat;
                        _e = (_d = JSON).stringify;
                        return [4 /*yield*/, browser.getTimeouts()];
                    case 4:
                        _b.apply(_a, [_c.apply("browser timeouts are ", [_e.apply(_d, [_j.sent(), null, 2])])]);
                        _j.label = 5;
                    case 5:
                        if (!(browser instanceof multiremote_1.MultiRemoteDriver)) return [3 /*break*/, 17];
                        _j.label = 6;
                    case 6:
                        _j.trys.push([6, 14, 15, 16]);
                        _f = __values(browser.instances), _g = _f.next();
                        _j.label = 7;
                    case 7:
                        if (!!_g.done) return [3 /*break*/, 13];
                        name_1 = _g.value;
                        if (!this._capabilities[name_1].capabilities["wdi5:authentication"]) return [3 /*break*/, 9];
                        return [4 /*yield*/, (0, wdi5_bridge_1.authenticate)(this._capabilities[name_1].capabilities["wdi5:authentication"], name_1)];
                    case 8:
                        _j.sent();
                        _j.label = 9;
                    case 9:
                        if (!!this._config.wdi5.skipInjectUI5OnStart) return [3 /*break*/, 11];
                        return [4 /*yield*/, this.injectUI5(browser[name_1])];
                    case 10:
                        _j.sent();
                        return [3 /*break*/, 12];
                    case 11:
                        Logger.warn("skipped wdi5 injection!");
                        _j.label = 12;
                    case 12:
                        _g = _f.next();
                        return [3 /*break*/, 7];
                    case 13: return [3 /*break*/, 16];
                    case 14:
                        e_1_1 = _j.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 16];
                    case 15:
                        try {
                            if (_g && !_g.done && (_h = _f["return"])) _h.call(_f);
                        }
                        finally { if (e_1) throw e_1.error; }
                        return [7 /*endfinally*/];
                    case 16: return [3 /*break*/, 22];
                    case 17:
                        if (!this._capabilities["wdi5:authentication"]) return [3 /*break*/, 19];
                        return [4 /*yield*/, (0, wdi5_bridge_1.authenticate)(this._capabilities["wdi5:authentication"])];
                    case 18:
                        _j.sent();
                        _j.label = 19;
                    case 19:
                        if (!!this._config.wdi5.skipInjectUI5OnStart) return [3 /*break*/, 21];
                        return [4 /*yield*/, this.injectUI5(browser)];
                    case 20:
                        _j.sent();
                        return [3 /*break*/, 22];
                    case 21:
                        Logger.warn("skipped wdi5 injection!");
                        _j.label = 22;
                    case 22: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * this is a helper function to late-inject ui5 at test-time
     * it relays the the wdio configuration (set in the .before() hook to the browser.config parameter by wdio)
     * to the injectUI5 function of the actual wdi5-bridge
     */
    Service.prototype.injectUI5 = function (browserInstance) {
        if (browserInstance === void 0) { browserInstance = browser; }
        return __awaiter(this, void 0, void 0, function () {
            var config;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, wdi5_bridge_1.checkForUI5Page)(browserInstance)];
                    case 1:
                        if (!_a.sent()) return [3 /*break*/, 3];
                        config = this._config ? this._config : browserInstance.config;
                        return [4 /*yield*/, (0, wdi5_bridge_1.injectUI5)(config, browserInstance)];
                    case 2:
                        _a.sent();
                        return [3 /*break*/, 4];
                    case 3: throw new Error("wdi5: no UI5 page/app present to work on :(");
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    return Service;
}());
exports["default"] = Service;
